File: uncategorized/cpp/strategy.cpp
#include <stdio.h>
#include <stdlib.h>

void solve(int n0, int c1, int c2, int c3, int time);
int t,n,max,a[2][11];

main(){
       scanf("%d",&t);
       for (int kasus=0; kasus<t; kasus++){
           scanf("%d",&n);
           for (int i=0; i<3; i++)
               for (int j=0; j<n; j++)
                   scanf("%d",&a[i][j]);
           max=0;
           solve(0,0,0,0,0);
           printf("%d\n",max);
           }
}

void solve(int n0, int c1, int c2, int c3, int time){
     if ((n0==n) or (time==280))
        if (c1+c2+c3>max) max=c1+c2+c3;
           else {}
        else {
             n0++;
             solve(n0,c1,c2,c3,time);
             if ((c1<(n+1)/2)/* and ((time+a[0,n0])<=280)*/)
                solve(n0,c1,c2,c3,a[0,n0]);
             
             }
}

File: uncategorized/pascal/strategy.pas
type
  con=array[1..4] of integer;

var
  t,kasus,n,max,mid,k,i:integer;
  c:con;
  a:array[1..12] of con;

procedure solve(n0,time:integer; c:con);
  var
    i:integer;

  begin
    if (n0=n) or (time=280) then
      if c[1]+c[2]+c[3]>max then max:=c[1]+c[2]+c[3]
        else
      else
        begin
          solve(n0+1,time,c);
          for i:=1 to 3 do
            if (time+a[n0+1,i]<=280) and (c[i]<mid) then
              begin
                inc(c[i]);
                solve(n0+1,time+a[n0+1,i],c);
                dec(c[i]);
              end;
        end;
  end;

begin
  readln(t);
  for kasus:=1 to t do
    begin
      readln(n);
      for k:=1 to 3 do
        for i:=1 to n do read(a[i][k]);
      max:=0;
      mid:=(n+1) div 2;
      solve(0,0,c);
      writeln(max);
    end;
end.
